src/cmp.rs

```diff
@@ -0,0 +1,267 @@
+//! Diff 2 icon fonts to find out what got added,modified,removed.
+//!
+
+use crate::{
+    error::IconResolutionError,
+    iconid::{get_icons, Icon},
+    pens::SvgPathPen,
+};
+use core::cmp::PartialEq;
+use kurbo::BezPath;
+use rayon::prelude::*;
+use skrifa::{
+    instance::{Location, Size},
+    outline::DrawSettings,
+    raw::{tables::gvar::Gvar, FontRef, ReadError, TableProvider},
+    GlyphId, MetadataProvider, OutlineGlyph, OutlineGlyphCollection,
+};
+use std::collections::HashMap;
+
+#[derive(Debug, PartialEq)]
+pub struct CompareResult {
+    pub added: Vec<String>,
+    pub modified: Vec<String>,
+    pub removed: Vec<String>,
+}
+
+///
+/// Compares 2 icon fonts `lhs`, `rhs` and returns names for
+/// `CompareResult::added`: icons in `lhs` but not in `rhs`.
+/// `CompareResult::modified`: icons in both `lhs` and `rhs` that draws differently.
+/// `CompareResult::removed`: icons not in `lhs` but in `rhs`.
+///
+pub fn cmp(lhs: &FontRef, rhs: &FontRef) -> Result<CompareResult, IconResolutionError> {
+    let lhs_icons = get_icons(lhs)?;
+    let rhs_icons = get_icons(rhs)?;
+    let lhs_icons: HashMap<String, GlyphId> = map_by_names(lhs_icons);
+    let rhs_icons: HashMap<String, GlyphId> = map_by_names(rhs_icons);
+    let added = in_first_but_not_second(&lhs_icons, &rhs_icons);
+    let removed = in_first_but_not_second(&rhs_icons, &lhs_icons);
+    let modified = diff_glyphs(lhs_icons, rhs_icons, lhs, rhs)?;
+    Ok(CompareResult {
+        added,
+        modified,
+        removed,
+    })
+}
+
+fn diff_glyphs(
+    lhs_icons: HashMap<String, GlyphId>,
+    rhs_icons: HashMap<String, GlyphId>,
+    lhs: &FontRef,
+    rhs: &FontRef,
+) -> Result<Vec<String>, IconResolutionError> {
+    let lhs_outlines = Tables::new(lhs)?;
+    let rhs_outlines = Tables::new(rhs)?;
+    let common: Vec<(String, GlyphId, GlyphId)> = lhs_icons
+        .into_iter()
+        .filter_map(|(k, v)| rhs_icons.get(&k).map(|r_gid| (k, v, *r_gid)))
+        .collect();
+    let result = common
+        .par_iter()
+        .map(|(name, lhs_gid, rhs_gid)| {
+            let mut lhs_closure: Vec<_> = lhs
+                .gsub()?
+                .closure_glyphs([*lhs_gid].into())?
+                .into_iter()
+                .collect();
+            let mut rhs_closure: Vec<_> = rhs
+                .gsub()?
+                .closure_glyphs([*rhs_gid].into())?
+                .into_iter()
+                .collect();
+            if lhs_closure.len() != rhs_closure.len() {
+                return Ok::<String, IconResolutionError>(name.to_string());
+            }
+            lhs_closure.sort();
+            rhs_closure.sort();
+            for (lhs_gid, rhs_gid) in lhs_closure.iter().zip(rhs_closure.iter()) {
+                if !eq(&lhs_outlines, &rhs_outlines, *lhs_gid, *rhs_gid)? {
+                    return Ok(name.to_string());
+                }
+            }
+            Ok(String::from(""))
+        })
```

Comment:
I understand what's going on here but it's a bit difficult to untangle without reading through it a few times. I think the issue is using `Result<String, _>` to cover three cases (error, same, different) with the empty string being the sentinel for "same." I'd suggest replacing this with `Result<Option<String>>, _>` with `Ok(None)` representing "same" and also adding a couple comments describing what the code is doing for the sake of future readers.